src/storage-controller/src/lib.rs

```diff
@@ -634,76 +580,60 @@ where
-                        // as far the dependent's upper.
-                        mz_ore::soft_assert_or_log!(
-                            write_frontier.elements() == &[T::minimum()]
-                                || PartialOrder::less_than(&dependency_since, write_frontier),
-                            "dependency ({dep}) since has advanced past dependent ({id}) upper \n
-                            dependent ({id}): since {:?}, upper {:?} \n
-                            dependency ({dep}): since {:?}",
-                            data_shard_since,
-                            write_frontier,
-                            dependency_since
-                        );
+            let storage_dependencies =
+                self.determine_collection_dependencies(id, &description.data_source)?;
 
-                        dependency_since
-                    } else {
-                        data_shard_since
-                    }
-                }
-                None => data_shard_since,
-            };
+            let dependency_read_holds = self
+                .storage_collections
+                .acquire_read_holds(storage_dependencies)
+                .expect("can acquire read holds");
 
-            let collection_state = CollectionState::new(
-                description,
-                initial_since,
-                write_frontier.clone(),
-                storage_dependency,
-                metadata.clone(),
-            );
+            let mut dependency_since = Antichain::from_elem(T::minimum());
+            for read_hold in dependency_read_holds.iter() {
+                dependency_since.join_assign(read_hold.since());
+            }
+
+            // Assert some invariants.
+            if !dependency_read_holds.is_empty() {
+                // The dependency since cannot be in advance of the
+                // dependent upper unless the collection is new. If the
+                // dependency since advanced past the dependent's upper,
+                // the dependent cannot read data from the dependency at
+                // its upper.
+                //
+                // Another way of understanding that this is a problem
+                // is that this means that the read hold installed on
+                // the dependency was probably not been upheld––if it
+                // were, the dependency's since could not have advanced
+                // as far the dependent's upper.
+                mz_ore::soft_assert_or_log!(
+                    write_frontier.elements() == &[T::minimum()]
+                        || PartialOrder::less_than(&dependency_since, write_frontier),
+                    "dependency since has advanced past dependent ({id}) upper \n
+                            dependent ({id}): upper {:?} \n
+                            dependency since {:?}",
+                    write_frontier,
+                    dependency_since
+                );
+            }
```

Comment:
I tried making the two as independent as possible. So it's also checking the invariants here. Because in the future `StorageCollections` will not be an implementation detail of `StorageController` but be more like a peer, in the controller scene.

Reply:
that its an iterative change for now makes sense!